src/Symfony/Component/Security/Http/Controller/SecurityTokenValueResolver.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class SecurityTokenValueResolver implements ValueResolverInterface
+{
+    public function __construct(private readonly TokenStorageInterface $tokenStorage)
+    {
+    }
+
+    /**
+     * @return TokenInterface[]
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument): array
+    {
+        if (!($type = $argument->getType()) || !is_subclass_of($type, TokenInterface::class)) {
+            return [];
+        }
+
+        if (null === $token = $this->tokenStorage->getToken()) {
+            if ($argument->isNullable()) {
+                return [];
+            }
+
+            throw new HttpException(Response::HTTP_UNAUTHORIZED, 'The security token required but not provided.');
```

Comment:
`AccessDeniedException` v `AccessDeniedHttpException` v `HttpException` keeps confusing me :sweat_smile:

Reply:
Yeah, I know. But semantically 401 is more logical, isn't?

Token not provided - 401
Token provided but without required role - 403